### PR TITLE
TagBox should wait for data loading to change selection (close T673636)

### PR DESCRIPTION
**Pull request by churkin**

### CODE REVIEW

File: js/ui/tag_box.js [hunk 1]
```diff
@@ -729,14 +729,21 @@ var TagBox = SelectBox.inherit({
...
         this._$tagsContainer = this.$element()
             .find("." + TEXTEDITOR_CONTAINER_CLASS)
             .addClass(TAGBOX_TAG_CONTAINER_CLASS)
             .addClass(NATIVE_CLICK_CLASS);
 
         this._renderInputSize();
-        this._renderTags();
-        this._popup && this._popup.refreshPosition();
+        this._renderTags().always((function() {
+            that._popup && that._popup.refreshPosition();
```
Comment (Yarovaya) [inline js/ui/tag_box.js]: You can use this

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 2]
```diff
@@ -4832,6 +4832,30 @@ QUnit.module("regression", {
     }
 });
 
+QUnit.test("Selection refreshing process should wait for the items data will be loaded from the data source (T673636)", function(assert) {
+    var clock = sinon.useFakeTimers(),
+        refreshSelectedCalled = false,
+        tagBox = $("#tagBox").dxTagBox({
+            valueExpr: "id",
+            dataSource: {
+                load: () => {
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js:4840]: Don't use both ES5 and ES6

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 3]
```diff
@@ -4832,6 +4832,30 @@ QUnit.module("regression", {
...
+                    var d = $.Deferred();
+
+                    setTimeout(() => d.resolve([]), 0);
+
+                    return d.promise();
+                }
+            }
+        }).dxTagBox("instance");
+
+    tagBox._refreshSelected = () => refreshSelectedCalled = true;
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Don't use private methods in the test

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 4]
```diff
@@ -4832,6 +4832,30 @@ QUnit.module("regression", {
     }
 });
 
+QUnit.test("Selection refreshing process should wait for the items data will be loaded from the data source (T673636)", function(assert) {
+    var clock = sinon.useFakeTimers(),
+        refreshSelectedCalled = false,
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Use sinon.stub or sinon.spy for this purpose

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js [hunk 5]
```diff
@@ -4832,6 +4832,30 @@ QUnit.module("regression", {
     }
 });
 
+QUnit.test("Selection refreshing process should wait for the items data will be loaded from the data source (T673636)", function(assert) {
+    var clock = sinon.useFakeTimers(),
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js:4836]: Restore the clock after the test or use this.clock

**Review by mrjohn0011 — CHANGES_REQUESTED** [review]
Test should be refactored

File: testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js (hunk 3 above)
Comment (churkin) [reply testing/tests/DevExpress.ui.widgets.editors/tagBox.tests.js]: Any suggestions?

**Review by Yarovaya — DISMISSED** [review]

**Review by mrjohn0011 — APPROVED** [review]

**Review by Yarovaya — APPROVED** [review]

**Pull request merged**